File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Correctness/CorrectnessHarness.cs
namespace Ecommerce.Correctness;

/// <summary>
/// Intentional static call graph for rgctl expected-facts checks.
/// Prefer static calls so extraction does not depend on DI.
/// </summary>
public static class CorrectnessHarness
{
    /// <summary>Leaf — no outbound application calls.</summary>
    public static int CorrectnessLeaf() => 42;

    /// <summary>Mid — calls <see cref="CorrectnessLeaf"/>.</summary>
    public static int CorrectnessMid() => CorrectnessLeaf() + 1;

    /// <summary>Root — calls <see cref="CorrectnessMid"/> and branches for CFG.</summary>
    public static int CorrectnessRoot(bool flag)
    {
        var value = CorrectnessMid();
        if (flag)
        {
            return value * 2;
        }
        return value;
    }
}

File: rgctl-tests/ecommerce-java/src/main/java/com/example/ecommerce/correctness/CorrectnessHarness.java
package com.example.ecommerce.correctness;

/**
 * Intentional static call graph for rgctl expected-facts checks.
 *
 * <p>Keep method names unique and stable. Prefer static calls so extraction does not
 * depend on Spring DI field resolution.
 */
public final class CorrectnessHarness {

    private CorrectnessHarness() {}

    /** Leaf — no outbound application calls. */
    public static int correctnessLeaf() {
        return 42;
    }

    /** Mid — calls {@link #correctnessLeaf()}. */
    public static int correctnessMid() {
        return correctnessLeaf() + 1;
    }

    /**
     * Root — calls {@link #correctnessMid()} and branches for a non-trivial CFG.
     *
     * @param flag branch selector
     * @return transformed mid value
     */
    public static int correctnessRoot(boolean flag) {
        int value = correctnessMid();
        if (flag) {
            return value * 2;
        }
        return value;
    }

    /** Shared sink for diamond topology QE. */
    public static int correctnessShared() {
        return 1;
    }

    public static int correctnessLeft() {
        return correctnessShared();
    }

    public static int correctnessRight() {
        return correctnessShared();
    }

    /** Diamond root — dual callers into {@link #correctnessShared()}. */
    public static int correctnessDiamond() {
        return correctnessLeft() + correctnessRight();
    }
}
